torch/csrc/jit/codegen/cuda/lower_vectorize_welford.cpp

```diff
@@ -0,0 +1,567 @@
+    // The thread predicate may be different from the other
+    // expressions in the same innermost loop. Only allows if it's not
+    // used or known to be true. We could check if all the other
+    // expressions have the same thread predicate, though not sure if
+    // it could have any meaningful benfit
+    auto thread_pred = wop_ite->predicate()->thread_pred();
+    if (thread_pred != nullptr &&
+        (!thread_pred->isConst() || !thread_pred->value())) {
+      return false;
+    }
+
+    TORCH_INTERNAL_ASSERT(!for_loops_.empty());
+    auto innermost_loop = for_loops_.back();
+
+    // Check all the exprs in the same scope
+    for (auto expr : innermost_loop->body().exprs()) {
+      // Bail out if a loop is found
+      if (expr->isA<kir::ForLoop>()) {
+        return false;
+      }
+
+      // If expr is not an IfThenElse and not predicated, nothing to
+      // worry about
+      if (!expr->isA<kir::IfThenElse>() && expr->predicate() == nullptr) {
+        continue;
+      }
+
+      // Consider only when this is a predicated expr, i.e, expr is an
+      // IfThenElse containing just one expr. When it's not an
+      // IfThenElse, it should be an expr with predicates embedded in
+      // it as predicate_, e.g., block/grid reductions. It is highly
+      // unlikely such expressions show up in the same ForLoop as a
+      // sequential WelfordOp, so not handling them shouldn't be a
+      // problem.
+      auto expr_ite = dynamic_cast<kir::IfThenElse*>(expr);
+      if (expr_ite == nullptr ||
+          !(expr_ite->thenBody().size() == 1 &&
+            expr_ite->elseBody().size() == 0)) {
+        return false;
+      }
+
+      expr = expr_ite->thenBody().at(0);
+
+      // Check only other expressions than the WelfordOp itself
+      if (expr == wop) {
+        continue;
+      }
+
+      // Only TV ops should appear
+      if (!ir_utils::isTvOp(expr)) {
+        return false;
+      }
+
+      TensorView* tv = ir_utils::getTvOutput(expr);
+
+      // tv should never be nullptr as non-TvOp is filtered out, but
+      // if that happens, just bails out
+      if (tv == nullptr) {
+        return false;
+      }
+
+      // Make sure this expr doesn't have a thread predicate
+      auto expr_thread_pred = expr_ite->predicate()->thread_pred();
+      if (expr_thread_pred != nullptr &&
+          (!expr_thread_pred->isConst() || !expr_thread_pred->value())) {
+        return false;
+      }
+
+      // Check if the predicate of the expr is known to be true
+      if (expr_ite->predicate()->isConst() &&
+          expr_ite->predicate()->value()->value()) {
+        continue;
+      }
+
+      // If tv is fully loop-mapped with the welford output TV, it is
+      // guaranteed that its predicate can be safely replaced with the
+      // predicate of the WelfordOp. If not, that doesn't necessarily
+      // mean the expr predicate is different, but likely not
+      // worthwhile to consider.
```

Comment:
If two expressions are loop mapped, it doesn't necessarily mean they have the same predicate, it just means their predicates are safe to join. This is because of broadcast resolution, the full predicate is a composite of all the root domains covered in the roots of the expressions. I don't see any specific errors in the code, I'm just stating this is how we do the unrolled predicate, accumulate all the predicates from the various expressions to make sure we covered all the root IterDomains.